src/sqlfluff/rules/L042.py

```diff
@@ -48,44 +70,281 @@ class Rule_L042(BaseRule):
+
+
+def _calculate_fixes(
+    dialect: Dialect,
+    root_select: Segments,
+    nested_subqueries: List[Tuple[str, Segments, Segments, BaseSegment]],
+):
+    """Given the Root select and the offending subqueries calculate fixes."""
+    Seg = partial(_get_seg, dialect=dialect)
+    ctes = _CTEChecker()
+    is_with = root_select.all(is_type("with_compound_statement"))
+    # TODO: consider if we can fix recursive CTEs
+    is_recursive = is_with and len(root_select.children(is_name("recursive"))) > 0
+    segmentify = partial(
+        _segmentify,
+        casing=_get_casing_preference(root_select),
+    )
+    # Init the output/final select &
+    # populate existing CTEs
+    output_select = root_select
+    if is_with:
+        output_select = root_select.children(
+            is_type(
+                "set_expression",
+                "select_statement",
+            )
+        )
+        for cte in root_select.children(is_type("common_table_expression")):
+            assert isinstance(cte, CTEDefinitionSegment), "TypeGaurd"
+            ctes.insert_cte(cte)
+
+    lint_results: List[LintResult] = []
+    CTESegment = Seg(CTEDefinitionSegment)
+    TableExpressionSeg = Seg(TableExpressionSegment)
+    TableReferenceSeg = Seg(TableReferenceSegment)
+    for parent_type, _, this_seg, subquery in nested_subqueries:
+        alias_name = ctes.get_name(this_seg.children(is_type("alias_expression")))
+        ctes.insert_cte(
+            CTESegment(
+                segments=(
+                    CodeSegment(
+                        raw=alias_name,
+                        name="naked_identifier",
+                        type="identifier",
+                    ),
+                    WhitespaceSegment(),
+                    segmentify("AS"),
+                    WhitespaceSegment(),
+                    subquery,
                 )
-                if bracketed_expression:
-                    from_expression_element = bracketed_expression
-
-                # If we find a child with a "problem" type, raise an issue.
-                # If not, we're fine.
-                seg = from_expression_element.children(
-                    is_type(
-                        "with_compound_statement",
-                        "set_expression",
-                        "select_statement",
-                    )
+            )
+        )
+        # TODO: Create non-mutative helper function.
+        # We will replace the whole tree, mutate the table expression.
+        this_seg[0].segments = (
+            TableExpressionSeg(
+                segments=(
+                    TableReferenceSeg(
+                        segments=(
+                            CodeSegment(
+                                raw=alias_name,
+                                name="naked_identifier",
+                                type="identifier",
+                            ),
+                        )
+                    ),
                 )
```

Comment:
Without all this nesting I was corrupting the parse tree - PS nice catch with the new tests @barrywhart

Reply:
I have spent a **lot** of time staring at `sqlfluff parse` output the past few weeks!!